New dependency for building/testing: docker?

I do have the impression that you need docker in order to build/test recent versions of gitbucket.
I guess this should be documented within README.md if it is going to stay this way...

```
$ sbt test
...
[info] visibleFor
[info] - should return true for visible pagination links
[info] GitBucketCoreModuleSpec:
[info] - Migration H2
06:55:23.492 [pool-1-thread-1-ScalaTest-running-GitBucketCoreModuleSpec] INFO  o.t.d.DockerClientProviderStrategy - Will use 'okhttp' transport
06:55:33.943 [pool-1-thread-1-ScalaTest-running-GitBucketCoreModuleSpec] ERROR o.t.d.EnvironmentAndSystemPropertyClientProviderStrategy - ping failed with configuration Environment variables, system properties and defaults. Resolved: 
    dockerHost=unix:///var/run/docker.sock
    apiVersion='{UNKNOWN_VERSION}'
    registryUrl='https://index.docker.io/v1/'
    registryUsername='uli'
    registryPassword='null'
    registryEmail='null'
    dockerConfig='DefaultDockerClientConfig[dockerHost=unix:///var/run/docker.sock,registryUsername=uli,registryPassword=<null>,registryEmail=<null>,registryUrl=https://index.docker.io/v1/,dockerConfigPath=/home/uli/.docker,sslConfig=<null>,apiVersion={UNKNOWN_VERSION},dockerConfig=<null>]'
 due to org.rnorth.ducttape.TimeoutException: Timeout waiting for result with exception
org.rnorth.ducttape.TimeoutException: Timeout waiting for result with exception
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:51)
	at org.testcontainers.dockerclient.DockerClientProviderStrategy.ping(DockerClientProviderStrategy.java:190)
	at org.testcontainers.dockerclient.EnvironmentAndSystemPropertyClientProviderStrategy.test(EnvironmentAndSystemPropertyClientProviderStrategy.java:42)
	at org.testcontainers.dockerclient.DockerClientProviderStrategy.lambda$getFirstValidStrategy$2(DockerClientProviderStrategy.java:113)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:267)
	at java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(StreamSpliterators.java:302)
	at java.util.stream.Streams$ConcatSpliterator.tryAdvance(Streams.java:731)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findAny(ReferencePipeline.java:469)
	at org.testcontainers.dockerclient.DockerClientProviderStrategy.getFirstValidStrategy(DockerClientProviderStrategy.java:148)
	at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:105)
	at org.testcontainers.containers.GenericContainer.<init>(GenericContainer.java:142)
	at org.testcontainers.containers.JdbcDatabaseContainer.<init>(JdbcDatabaseContainer.java:45)
	at org.testcontainers.containers.MySQLContainer.<init>(MySQLContainer.java:29)
	at com.dimafeng.testcontainers.MySQLContainer.$anonfun$container$2(MySQLContainer.scala:18)
	at scala.Option.getOrElse(Option.scala:138)
	at com.dimafeng.testcontainers.MySQLContainer.<init>(MySQLContainer.scala:18)
	at gitbucket.core.GitBucketCoreModuleSpec$$anon$1.<init>(GitBucketCoreModuleSpec.scala:29)
	at gitbucket.core.GitBucketCoreModuleSpec.$anonfun$new$3(GitBucketCoreModuleSpec.scala:29)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
	at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
	at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
	at org.scalatest.FunSuite.withFixture(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.invokeWithFixture$1(FunSuiteLike.scala:184)
	at org.scalatest.FunSuiteLike.$anonfun$runTest$1(FunSuiteLike.scala:196)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
	at org.scalatest.FunSuiteLike.runTest(FunSuiteLike.scala:196)
	at org.scalatest.FunSuiteLike.runTest$(FunSuiteLike.scala:178)
	at org.scalatest.FunSuite.runTest(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.$anonfun$runTests$1(FunSuiteLike.scala:229)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:396)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:379)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
	at org.scalatest.FunSuiteLike.runTests(FunSuiteLike.scala:229)
	at org.scalatest.FunSuiteLike.runTests$(FunSuiteLike.scala:228)
	at org.scalatest.FunSuite.runTests(FunSuite.scala:1560)
	at org.scalatest.Suite.run(Suite.scala:1147)
	at org.scalatest.Suite.run$(Suite.scala:1129)
	at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.$anonfun$run$1(FunSuiteLike.scala:233)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
	at org.scalatest.FunSuiteLike.run(FunSuiteLike.scala:233)
	at org.scalatest.FunSuiteLike.run$(FunSuiteLike.scala:232)
	at org.scalatest.FunSuite.run(FunSuite.scala:1560)
	at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:314)
	at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:507)
	at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:304)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: com.sun.jna.LastErrorException: [2] Datei oder Verzeichnis nicht gefunden
	at org.testcontainers.shaded.org.scalasbt.ipcsocket.UnixDomainSocket.<init>(UnixDomainSocket.java:62)
	at org.testcontainers.dockerclient.transport.okhttp.UnixSocketFactory$1.<init>(UnixSocketFactory.java:27)
	at org.testcontainers.dockerclient.transport.okhttp.UnixSocketFactory.createSocket(UnixSocketFactory.java:27)
	at org.testcontainers.shaded.okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:234)
	at org.testcontainers.shaded.okhttp3.internal.connection.RealConnection.connect(RealConnection.java:160)
	at org.testcontainers.shaded.okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257)
	at org.testcontainers.shaded.okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
	at org.testcontainers.shaded.okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
	at org.testcontainers.shaded.okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at org.testcontainers.shaded.okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at org.testcontainers.shaded.okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at org.testcontainers.shaded.okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at org.testcontainers.shaded.okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at org.testcontainers.shaded.okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at org.testcontainers.shaded.okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at org.testcontainers.shaded.okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at org.testcontainers.shaded.okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
	at org.testcontainers.shaded.okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at org.testcontainers.shaded.okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at org.testcontainers.shaded.okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at org.testcontainers.shaded.okhttp3.RealCall.execute(RealCall.java:77)
	at org.testcontainers.dockerclient.transport.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:262)
	at org.testcontainers.dockerclient.transport.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:257)
	at org.testcontainers.dockerclient.transport.okhttp.OkHttpInvocationBuilder.get(OkHttpInvocationBuilder.java:223)
	at org.testcontainers.dockerclient.transport.okhttp.OkHttpDockerCmdExecFactory$1.execute(OkHttpDockerCmdExecFactory.java:125)
	at org.testcontainers.dockerclient.transport.okhttp.OkHttpDockerCmdExecFactory$1.execute(OkHttpDockerCmdExecFactory.java:118)
	at com.github.dockerjava.core.exec.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:21)
	at com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:35)
	at org.testcontainers.dockerclient.DockerClientProviderStrategy.lambda$null$4(DockerClientProviderStrategy.java:193)
	at org.rnorth.ducttape.ratelimits.RateLimiter.getWhenReady(RateLimiter.java:51)
	at org.testcontainers.dockerclient.DockerClientProviderStrategy.lambda$ping$5(DockerClientProviderStrategy.java:191)
	at org.rnorth.ducttape.unreliables.Unreliables.lambda$retryUntilSuccess$0(Unreliables.java:41)
	... 4 common frames omitted
Caused by: com.sun.jna.LastErrorException: [2] Datei oder Verzeichnis nicht gefunden
	at org.testcontainers.shaded.org.scalasbt.ipcsocket.UnixDomainSocketLibrary.connect(Native Method)
	at org.testcontainers.shaded.org.scalasbt.ipcsocket.UnixDomainSocket.<init>(UnixDomainSocket.java:57)
	... 35 common frames omitted
06:55:33.952 [pool-1-thread-1-ScalaTest-running-GitBucketCoreModuleSpec] ERROR o.t.d.DockerClientProviderStrategy - Could not find a valid Docker environment. Please check configuration. Attempted configurations were:
06:55:33.952 [pool-1-thread-1-ScalaTest-running-GitBucketCoreModuleSpec] ERROR o.t.d.DockerClientProviderStrategy -     EnvironmentAndSystemPropertyClientProviderStrategy: failed with exception InvalidConfigurationException (ping failed)
06:55:33.952 [pool-1-thread-1-ScalaTest-running-GitBucketCoreModuleSpec] ERROR o.t.d.DockerClientProviderStrategy -     UnixSocketClientProviderStrategy: failed with exception InvalidConfigurationException (ping failed). Root cause NoSuchFileException (/var/run/docker.sock)
06:55:33.952 [pool-1-thread-1-ScalaTest-running-GitBucketCoreModuleSpec] ERROR o.t.d.DockerClientProviderStrategy - As no valid configuration was found, execution cannot continue
[info] - Migration MySQL 8.0 *** FAILED ***
[info]   java.lang.IllegalStateException: Could not find a valid Docker environment. Please see logs and check configuration
[info]   at org.testcontainers.dockerclient.DockerClientProviderStrategy.lambda$getFirstValidStrategy$3(DockerClientProviderStrategy.java:157)
[info]   at java.util.Optional.orElseThrow(Optional.java:290)
[info]   at org.testcontainers.dockerclient.DockerClientProviderStrategy.getFirstValidStrategy(DockerClientProviderStrategy.java:149)
[info]   at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:105)
[info]   at org.testcontainers.containers.GenericContainer.<init>(GenericContainer.java:142)
[info]   at org.testcontainers.containers.JdbcDatabaseContainer.<init>(JdbcDatabaseContainer.java:45)
[info]   at org.testcontainers.containers.MySQLContainer.<init>(MySQLContainer.java:29)
[info]   at com.dimafeng.testcontainers.MySQLContainer.$anonfun$container$2(MySQLContainer.scala:18)
[info]   at scala.Option.getOrElse(Option.scala:138)
[info]   at com.dimafeng.testcontainers.MySQLContainer.<init>(MySQLContainer.scala:18)
[info]   ...
[info] - Migration MySQL 5.7 *** FAILED ***
[info]   java.lang.IllegalStateException: Previous attempts to find a Docker environment failed. Will not retry. Please see logs and check configuration
[info]   at org.testcontainers.dockerclient.DockerClientProviderStrategy.getFirstValidStrategy(DockerClientProviderStrategy.java:78)
[info]   at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:105)
[info]   at org.testcontainers.containers.GenericContainer.<init>(GenericContainer.java:142)
[info]   at org.testcontainers.containers.JdbcDatabaseContainer.<init>(JdbcDatabaseContainer.java:45)
[info]   at org.testcontainers.containers.MySQLContainer.<init>(MySQLContainer.java:29)
[info]   at com.dimafeng.testcontainers.MySQLContainer.$anonfun$container$2(MySQLContainer.scala:18)
[info]   at scala.Option.getOrElse(Option.scala:138)
[info]   at com.dimafeng.testcontainers.MySQLContainer.<init>(MySQLContainer.scala:18)
[info]   at gitbucket.core.GitBucketCoreModuleSpec$$anon$1.<init>(GitBucketCoreModuleSpec.scala:29)
[info]   at gitbucket.core.GitBucketCoreModuleSpec.$anonfun$new$3(GitBucketCoreModuleSpec.scala:29)
[info]   ...
[info] - Migration PostgreSQL 11 *** FAILED ***
[info]   java.lang.IllegalStateException: Previous attempts to find a Docker environment failed. Will not retry. Please see logs and check configuration
[info]   at org.testcontainers.dockerclient.DockerClientProviderStrategy.getFirstValidStrategy(DockerClientProviderStrategy.java:78)
[info]   at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:105)
[info]   at org.testcontainers.containers.GenericContainer.<init>(GenericContainer.java:142)
[info]   at org.testcontainers.containers.JdbcDatabaseContainer.<init>(JdbcDatabaseContainer.java:45)
[info]   at org.testcontainers.containers.PostgreSQLContainer.<init>(PostgreSQLContainer.java:30)
[info]   at com.dimafeng.testcontainers.PostgreSQLContainer.<init>(PostgreSQLContainer.scala:12)
[info]   at com.dimafeng.testcontainers.PostgreSQLContainer$.apply(PostgreSQLContainer.scala:30)
[info]   at gitbucket.core.GitBucketCoreModuleSpec.$anonfun$new$5(GitBucketCoreModuleSpec.scala:52)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - Migration PostgreSQL 10 *** FAILED ***
[info]   java.lang.IllegalStateException: Previous attempts to find a Docker environment failed. Will not retry. Please see logs and check configuration
[info]   at org.testcontainers.dockerclient.DockerClientProviderStrategy.getFirstValidStrategy(DockerClientProviderStrategy.java:78)
[info]   at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:105)
[info]   at org.testcontainers.containers.GenericContainer.<init>(GenericContainer.java:142)
[info]   at org.testcontainers.containers.JdbcDatabaseContainer.<init>(JdbcDatabaseContainer.java:45)
[info]   at org.testcontainers.containers.PostgreSQLContainer.<init>(PostgreSQLContainer.java:30)
[info]   at com.dimafeng.testcontainers.PostgreSQLContainer.<init>(PostgreSQLContainer.scala:12)
[info]   at com.dimafeng.testcontainers.PostgreSQLContainer$.apply(PostgreSQLContainer.scala:30)
[info]   at gitbucket.core.GitBucketCoreModuleSpec.$anonfun$new$5(GitBucketCoreModuleSpec.scala:52)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] CommitStatusServiceSpec:
[info] - createCommitState can insert and update
[info] - getCommitStatus can find by commitId and context
[info] - getCommitStatus can find by commitStatusId
[info] AccountServiceSpec:
[info] - getAllUsers
[info] - getAccountByUserName
[info] - getAccountByMailAddress
[info] - updateLastLoginDate
[info] - updateAccount
[info] - group
[info] - createGroup save description
[info] - updateGroup save description
[info] LabelsServiceSpec:
[info] getLabels
[info] - should be empty when not have any labels
[info] - should return contained labels
[info] getLabel
[info] - should return None when the label not exist
[info] - should return a label fetched by label id
[info] - should return a label fetched by label name
[info] createLabel
[info] - should return accurate label id
[info] updateLabel
[info] - should change target label
[info] deleteLabel
[info] - should remove target label
[info] DirectorySpec:
[info] GitBucketHome
[info] - should set under target in test scope
[info] AvatarImageProviderSpec:
[info] getAvatarImageHtml
[info] - should show Gravatar image for no image account if gravatar integration is enabled
[info] - should show uploaded image even if gravatar integration is enabled
[info] - should show local image for no image account if gravatar integration is disabled
[info] - should show Gravatar image for specified mail address if gravatar integration is enabled
[info] - should show unknown image for unknown user if gravatar integration is enabled
[info] - should show unknown image for specified mail address if gravatar integration is disabled
[info] - should add tooltip if it's enabled
[info] StringUtilSpec:
[info] urlEncode
[info] - should encode whitespace to %20
[info] urlDecode
[info] - should decode encoded string to original string
[info] - should decode en%20 to whitespace
[info] splitWords
[info] - should split string by whitespaces
[info] escapeHtml
[info] - should escape &, <, > and "
[info] md5
[info] - should generate MD5 hash
[info] sha1
[info] - should generate SHA1 hash
[info] extractIssueId
[info] - should extract '#xxx' and return extracted id
[info] - should return Nil from message which does not contain #xxx
[info] extractCloseId
[info] - should extract 'close #xxx' and return extracted id
[info] - should returns Nil from message which does not contain close command
[info] getRepositoryViewerUrl
[info] - should convert GitBucket repository url
[info] - should convert GitHub repository url
[info] - should convert BitBucket repository url
[info] - should convert GitLab repository url
[info] MarkdownSpec:
[info] generateAnchorName
[info] - should convert whitespace characters to hyphens
[info] - should normalize characters with diacritics
[info] - should omit special characters
[info] escapeTaskList
[info] - should convert '- [ ] ' to '* task: :'
[info] - should convert '  - [ ] ' to '  * task: :'
[info] - should convert only first '- [ ] '
[info] - should convert '- [x] ' to '* task:x:'
[info] - should convert multi lines
[info] - should not convert if inserted before '- [ ] '
[info] - should not convert '- [] '
[info] - should not convert '- [ ]a'
[info] - should not convert '-[ ] '
[info] WebHookServiceSpec:
[info] - WebHookPullRequestService.getPullRequestsByRequestForWebhook
[info] - add and get and update and delete
[info] - getWebHooks, getWebHooksByEvent
[info] PullRequestServiceSpec:
[info] PullRequestService.getPullRequestFromBranch
[info] - should
[info] return pull request if exists pull request from `branch` to `defaultBranch` and not closed.
[info] return pull request if exists pull request from `branch` to other branch and not closed.
[info] return None if all pull request is closed
[info] EditorConfigUtilSpec:
[info] - no EditorConfig file
[info] - simple EditorConfig
[info] - .editorconfig parse error
[info] IssuesServiceSpec:
[info] - getCommitStatues
[info] Run completed in 1 minute, 10 seconds.
[info] Total number of tests run: 173
[info] Suites: completed 25, aborted 0
[info] Tests: succeeded 169, failed 4, canceled 0, ignored 0, pending 0
[info] *** 4 TESTS FAILED ***
[error] Failed tests:
[error] 	gitbucket.core.GitBucketCoreModuleSpec
[error] (Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 113 s, completed 14.01.2019 06:55:51
```

--- a/doc/build.md
+++ b/doc/build.md
@@ -53,3 +53,9 @@ To run the full series of tests, run the following command:
 ```shell
 $ sbt test
 ```
+
+If you don't have docker, you can skip docker tests which require docker as follows:
+
+```shell
+$ sbt "testOnly * -- -l ExternalDBTest"
+```

--- a/doc/build.md
+++ b/doc/build.md
@@ -3,7 +3,7 @@ How to build and run from the source tree
 
 First of all, Install [sbt](http://www.scala-sbt.org/index.html).
 
-```
+```shell
 $ brew install sbt
 ```
 
@@ -12,7 +12,7 @@ Run for Development
 
 If you want to test GitBucket, type the following command in the root directory of the source tree.
 
-```
+```shell
 $ sbt ~jetty:start
 ```
 
@@ -25,7 +25,7 @@ Build war file
 
 To build war file, run the following command:
 
-```
+```shell
 $ sbt package
 ```
 
@@ -33,7 +33,7 @@ $ sbt package
 
 To build an executable war file, run
 
-```
+```shell
 $ sbt executable
 ```
 
@@ -41,8 +41,15 @@ at the top of the source tree. It generates executable `gitbucket.war` into `tar
 
 Run tests spec
 ---------
+Before running tests, you need to install docker.
+
+```shell
+$ brew cask install docker       # Install Docker
+$ open /Applications/Docker.app  # Start Docker
+```
+
 To run the full series of tests, run the following command:
 
-```
+```shell
 $ sbt test
 ```

